python/tk_framework_desktopserver/shotgun/api_v2.py

```diff
@@ -0,0 +1,758 @@
+                            self.get_actions(data)
+                            return
+                    else:
+                        self.logger.debug("Commands not found in hash, caching now...")
+                        self._cache_actions(data, config_data)
+                        self.get_actions(data)
+                        return
+                except subprocess.CalledProcessError, e:
+                    self.logger.error(str(e))
+                    self.host.reply(
+                        dict(
+                            err="Shotgun Desktop failed to get engine commands.",
+                            retcode=self.CACHING_ERROR,
+                            out="Caching failed!",
+                        ),
+                    )
+                    return
+
+        self.host.reply(
+            dict(
+                err="",
+                retcode=self.SUCCESSFUL_LOOKUP,
+                actions=all_actions,
+                pcs=pc_names,
+            ),
+        )
+
+    ###########################################################################
+    # Context managers
+
+    @contextlib.contextmanager
+    def _db_connect(self):
+        """
+        Context manager that initializes a DB connection on enter and
+        disconnects on exit.
+        """
+        connection = self._init_db()
+        cursor = connection.cursor()
+        yield (connection, cursor)
+        cursor.close()
+        connection.close()
+
+    ###########################################################################
+    # Internal methods
+
+    def _cache_actions(self, data, config_data):
+        """
+        Triggers the caching or recaching of engine commands.
+
+        :param dict data: The data passed down from the wss client.
+        :param dict config_data: A dictionary keyed by PipelineConfiguration
+            id containing a dict that contains, at a minimum, "lookup_hash"
+            and "contents_hash" keys.
+        """
+        self.logger.debug("Caching engine commands...")
+
+        script = os.path.join(
+            os.path.dirname(__file__),
+            "scripts",
+            "cache_commands.py"
+        )
+
+        self.logger.debug("Executing script: %s" % script)
+        hash_data = dict()
+
+        for pc_id, pc_data in config_data.iteritems():
+            hash_data[pc_id] = dict(
+                lookup_hash=config_data[pc_id]["lookup_hash"],
+                contents_hash=config_data[pc_id]["contents_hash"],
+            )
+
+        args_file = self._get_arguments_file(
+            dict(
+                cache_file=self._cache_path,
+                data=data,
+                sys_path=sys.path,
+                base_configuration=self.BASE_CONFIG_URI,
+                hash_data=hash_data,
+                engine_name=self.ENGINE_NAME,
+            )
+        )
+
+        output = None
+
+        try:
+            output = subprocess.check_output([sys.executable, script, args_file])
```

Comment:
There's that, but there's also close_fds=True, which needs to be passed in. This is to avoid having child processes inherit the file descriptors from the parent process, which on Windows means that the port 9000 is considered in used as long as the desktop and any child process are running. The launch app on Windows actually blocks until the DCC is closed IIRC for certain DCCs IIRC, which caused this issue. At the very least I would recommend using Popen and doing something similar to [this](https://github.com/shotgunsoftware/tk-core/blob/develop/zero_config/python/tank/util/process.py#L57-L65) and pass in `close_fds=True`

Maybe we can start releasing with this bit only and see where that takes us.